src/client/render/plain.cpp

```diff
@@ -19,58 +19,141 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	context->hud->drawBlockBounds();
+	context->hud->drawSelectionMesh();
+	if (context->draw_wield_tool)
+		context->client->getCamera()->drawWieldedTool();
 }
 
-RenderingCorePlain::RenderingCorePlain(
-	IrrlichtDevice *_device, Client *_client, Hud *_hud)
-	: RenderingCore(_device, _client, _hud)
+void DrawHUD::run(PipelineContext *context)
 {
-	scale = g_settings->getU16("undersampling");
+	if (context->show_hud) {
+		if (context->shadow_renderer)
+			context->shadow_renderer->drawDebug();
+
+		if (context->draw_crosshair)
+			context->hud->drawCrosshair();
+
+		context->hud->drawHotbar(context->client->getEnv().getLocalPlayer()->getWieldIndex());
+		context->hud->drawLuaElements(context->client->getCamera()->getOffset());
+		context->client->getCamera()->drawNametags();
+		auto mapper = context->client->getMinimap();
+		if (mapper && context->show_minimap)
+			mapper->drawMinimap();
+	}
+	context->device->getGUIEnvironment()->drawAll();
 }
 
-void RenderingCorePlain::initTextures()
+
+void MapPostFxStep::setRenderTarget(RenderTarget * _target)
 {
-	if (scale <= 1)
-		return;
-	v2u32 size{scaledown(scale, screensize.X), scaledown(scale, screensize.Y)};
-	lowres = driver->addRenderTargetTexture(
-			size, "render_lowres", video::ECF_A8R8G8B8);
+	target = _target;
 }
 
-void RenderingCorePlain::clearTextures()
+void MapPostFxStep::run(PipelineContext *context)
 {
-	if (scale <= 1)
-		return;
-	driver->removeTexture(lowres);
+	if (target)
+		target->activate(context);
+
+	context->client->getEnv().getClientMap().renderPostFx(context->client->getCamera()->getCameraMode());
 }
 
-void RenderingCorePlain::beforeDraw()
+void RenderShadowMapStep::run(PipelineContext *context)
 {
-	if (scale <= 1)
-		return;
-	driver->setRenderTarget(lowres, true, true, skycolor);
+	// This is necessary to render shadows for animations correctly
+	context->device->getSceneManager()->getRootSceneNode()->OnAnimate(context->device->getTimer()->getTime());
+	context->shadow_renderer->update();
 }
 
-void RenderingCorePlain::upscale()
+// class UpscaleStep
+
+void UpscaleStep::run(PipelineContext *context)
 {
-	if (scale <= 1)
-		return;
-	driver->setRenderTarget(0, true, true);
-	v2u32 size{scaledown(scale, screensize.X), scaledown(scale, screensize.Y)};
-	v2u32 dest_size{scale * size.X, scale * size.Y};
-	driver->draw2DImage(lowres, core::rect<s32>(0, 0, dest_size.X, dest_size.Y),
-			core::rect<s32>(0, 0, size.X, size.Y));
+	video::ITexture *lowres = m_source->getTexture(0);
+	m_target->activate(context);
+	context->device->getVideoDriver()->draw2DImage(lowres,
+			core::rect<s32>(0, 0, context->target_size.X, context->target_size.Y),
+			core::rect<s32>(0, 0, lowres->getSize().Width, lowres->getSize().Height));
+}
+
+RenderStep *create3DStage()
+{
+	return new Draw3D();
+}
+
+RenderStep *create3DStage(Client *client, v2f scale)
+{
+	RenderStep *step = new Draw3D();
+	if (g_settings->getBool("enable_shaders")) {
+		RenderPipeline *pipeline = new RenderPipeline();
```

Comment:
I am somewhat confused about how many pipelines we have.
Initially I thought there is one. But then I see multiple places where we do `new RenderPipeLine()`.

Reply:
I use nested pipeline here to abstract the internals of 3D step+post processing from the stereo modes. I could squeeze it into a flat pipeline, but it would complicate the construction code.

Reply:
It's fine this way. I just was confused :)